### PR TITLE
add generator generic machine drivers

### PR DESCRIPTION
**Pull request by wlan0**
@cjellick @will-chan

### CODE REVIEW

File: generator/generator.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	_ "github.com/docker/machine/drivers/amazonec2"
+	//	_ "github.com/docker/machine/drivers/azure"
+	_ "github.com/docker/machine/drivers/digitalocean"
+	_ "github.com/docker/machine/drivers/exoscale"
+	//	_ "github.com/docker/machine/drivers/google"
```
Comment (wlan0) [inline generator/generator.go:14]: We are not doing google and azure now, because they require some UI and cattle changes

**Comment by will-chan** [comment]
General comment.  Can we keep the logging consistent?  We use WithFields(...) to display the variables rather than appending it to the string.

**Comment by wlan0** [comment]
@will-chan Its a lot of effort to change the logging, but provides a very small benefit . Moreover, this project is compiled and run separately from the rest of go-machine-service.

**Comment by cjellick** [comment]
I agree with @wlan0 that since this ran out-of-band from the actual service, it's not worth it. Moreover, one of the big benefits of of the WithFields() paradigm is that it makes the logs more parsible by programs (good for the service), but the down side is that it makes the logs a little harder for humans to read (bad for something that will be ran exclusively by a human).

File: generator/generator.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/codegangsta/cli"
+	"github.com/docker/machine/drivers"
+	"github.com/rancherio/go-machine-service/generator/helpers"
+)
+
+func main() {
+	err := generate()
+	if nil != err {
+		log.Fatal(err)
```
Comment (cjellick) [inline generator/generator.go:33]: can we exit with a non-zero code here?

Comment (wlan0) [inline generator/generator.go:33]: log.Fatal exits with 1

File: generator/generator.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/codegangsta/cli"
+	"github.com/docker/machine/drivers"
+	"github.com/rancherio/go-machine-service/generator/helpers"
+)
+
+func main() {
+	err := generate()
+	if nil != err {
```
Comment (cjellick) [inline generator/generator.go]: hm. Conventionally, it's `err != nil`, not the other way around. This bothers me.

File: generator/generator.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		log.Errorf("error generating field jsons err=%v", err)
+		return err
+	}
+	err = helpers.GenerateMachineJson(resourceData) // generate schema/base/machine.json.d/machine-generated.json
+	if nil != err {
+		log.Errorf("error generating machine json err=%v", err)
+		return err
+	}
+	return err
```
Comment (cjellick) [inline generator/generator.go]: Can you just `return nil` here to make it clear that if you get to this point in the code, you're going to have a clean exit?

File: generator/generator.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	if nil != err {
+		log.Fatal(err)
+	}
+}
+
+func generate() error {
+	resourceData, err := fetchData() // collects all data required to generate the files
+	if nil != err {
+		log.Errorf("error collecting driver data err=%v", err)
+		return err
```
Comment (cjellick) [inline generator/generator.go]: Logging and propagating an error is not a great convention. Could you construct a new error (using the message that you're currently logging) and just return that instead?

File: generator/generator.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	err = helpers.GenerateMachineJson(resourceData) // generate schema/base/machine.json.d/machine-generated.json
+	if nil != err {
+		log.Errorf("error generating machine json err=%v", err)
+		return err
+	}
+	return err
+}
+
+func loadBlacklist() []string {
+	blacklist := os.Getenv("RANCHER_DRIVER_BLACKLIST")
```
Comment (cjellick) [inline generator/generator.go:62]: I think a command line argument would make more sense than an environment variable here.

File: generator/generator.go (hunk 2 above)
Comment (cjellick) [inline generator/generator.go:33]: ah yeah. great.

File: generator/generator.go (hunk 6 above)
Comment (cjellick) [inline generator/generator.go:62]: discussed. blacklist not currently used. leaving as is.

File: generator/generator.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+			log.Errorf("error getting create flags for driver=%s, err=%v", driver, err)
+			return nil, err
+		}
+		for _, flagStruct := range createFlags {
+			var flagName, flagType string
+			var err error
+			switch flagStruct.(type) {
+			case cli.StringFlag:
+				flag := flagStruct.(cli.StringFlag)
+				flagName, err = getRancherName(flag.Name)
```
Comment (cjellick) [inline generator/generator.go:88]: Could this line be pulled out of each case and done just once at the top?

Comment (wlan0) [inline generator/generator.go:88]: It can't. Because the type of the struct is not known at that point. The interface{} cant access that field

**Comment by cjellick** [comment]
@wlan0 Can you figure out how to write some tests for this?

File: generator/generator.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	err = helpers.GenerateMachineJson(resourceData) // generate schema/base/machine.json.d/machine-generated.json
+	if err != nil {
+		return fmt.Errorf("error generating machine json err=%v", err)
+	}
+	return nil
+}
+
+func loadBlacklist() []string {
+	blacklist := os.Getenv("RANCHER_DRIVER_BLACKLIST")
+	if "" != blacklist {
```
Comment (cjellick) [inline generator/generator.go]: Can you change all comparisons to have the variable on the left?

Comment (wlan0) [inline generator/generator.go]: yes

**Comment by wlan0** [comment]
@cjellick - I discussed with @will-chan and we figured that the validation tests done by @sangeethah would suffice to validate that this is working.

File: generator/generator.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+	return &helpers.ResourceData{Blacklist: blacklist, Drivers: drivers.GetDriverNames(), ResourceMap: resourceMap}, nil
+}
+
+func getRancherName(machineFlagName string) (string, error) {
+	parts := strings.SplitN(machineFlagName, "-", 2)
+	if len(parts) != 2 {
+		errString := fmt.Sprintf("parameter %s does not follow expected naming convention [DRIVER]-[FLAG-NAME]", machineFlagName)
+		return "", fmt.Errorf(errString)
```
Comment (cjellick) [inline generator/generator.go]: No need to fmt.Sprintf and fmt.Errorf. just put the string and its parameters into Errorf

File: generator/generator.go (hunk 7 above)
Comment (cjellick) [inline generator/generator.go:88]: roger.

Comment (cjellick) [inline generator/generator.go:88]: Though, there does appear to be a GenericFlag interface that has a getName() method.

Comment (wlan0) [inline generator/generator.go:88]: GenericFlag is not the Flag interface{}. It is a type, just like IntFlag and BoolFlag.

Comment (cjellick) [inline generator/generator.go:88]: This thing:
https://github.com/codegangsta/cli/blob/master/flag.go#L34-L39
but whatever, its fine as is.

**Comment by cjellick** [comment]
I don't like having completely untested code. Even if it were just a smoke test that didn't assert much, but verified the thing could at least run, that would be nice

**Comment by wlan0** [comment]
@cjellick I understand. I'll add some tests

File: generator/helpers/spring.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+const bodyTemplate = `{{range .}}                <value>{{print .}}Config,parent=baseMachineConfig</value>  
+{{end}}`
+
+const header = `<beans xmlns="http://www.springframework.org/schema/beans"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:context="http://www.springframework.org/schema/context"
+    xmlns:aop="http://www.springframework.org/schema/aop"
+    xmlns:util="http://www.springframework.org/schema/util"
```
Comment (cjellick) [inline generator/helpers/spring.go:15]: Why have the aop and util namespaces in here?

File: generator/generator.go (hunk 7 above)
Comment (wlan0) [inline generator/generator.go:88]: That is an unexported method `getName()`

**Comment by cjellick** [comment]
Reviewing complete. other than a smoke test to prove it runs, looks good. let me know when u add the script

**Comment by cjellick** [comment]
Just noticed:
can you add teh generated directory to .gitignore:
generator/generated/

**Comment by cjellick** [comment]
@wlan0 Last thing: can you call this script from scripts/test so that it gets executed as part of the CI process? That's the main thing I wanted to get: proving that it works via CI.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**